crd: add a flag to specify the scope of the crd

xcoulon opened the pull request:
**Description of the change:**
- add a `--cluster-scoped-crd` flag in the `add api`
and `add crd` commands
- rename the `--cluster-scoped` flag to `-cluster-scoped-operator`
in the `new` command, and add a `--cluster-scoped-crd` flag, but
only for the `ansible` and `helm` types or operators


**Motivation for the change:**
See https://github.com/operator-framework/operator-sdk/issues/890#issuecomment-452809400


Fixes #890

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

estroz commented on commands/operator-sdk/cmd/new.go:
```diff
@@ -288,9 +292,9 @@ func doHelmScaffold() error {
+		&scaffold.Role{IsClusterScoped: isClusterScopedOperator},
+		&scaffold.RoleBinding{IsClusterScoped: isClusterScopedOperator},
+		&helm.Operator{IsClusterScoped: isClusterScopedOperator},
 		&scaffold.CRD{Resource: resource},
```
This also should receive the `isClusterScoped` variable correct?

xcoulon replied:
@estroz ah yes, indeed!

xcoulon replied:
fixed in 543b7d0

estroz commented on pkg/scaffold/crd.go:
```diff
@@ -145,25 +151,30 @@ func (s *CRD) CustomRender() ([]byte, error) {
 
-func newCRDForResource(r *Resource) *apiextv1beta1.CustomResourceDefinition {
-	return &apiextv1beta1.CustomResourceDefinition{
+func (s *CRD) newCRD() (*apiextv1beta1.CustomResourceDefinition, error) {
```
No need to return an error.

xcoulon replied:
right, this is a left-over form a previous idea, and I actually don't need it anymore

xcoulon replied:
fixed in 543b7d0

estroz commented on pkg/scaffold/operator.go:
```diff
@@ -24,7 +24,6 @@ const OperatorYamlFile = "operator.yaml"
 
 type Operator struct {
 	input.Input
-
```
Keep this line separator.

xcoulon replied:
ok

xcoulon replied:
fixed in 543b7d0

estroz commented on pkg/scaffold/scaffold.go:
```diff
@@ -181,7 +181,7 @@ func (s *Scaffold) doRender(i input.Input, e input.File, absPath string) error {
 	}
 	_, err = f.Write(b)
-	log.Infoln("Created", i.Path)
+	log.Infof("Created '%s' (%s)\n", i.Path, absPath)
```
Undo this change (outside of the scope of this PR).

xcoulon replied:
sure, let me revert it

xcoulon replied:
done in 1c31dac

estroz commented on test/e2e/doc.go:
```diff
@@ -0,0 +1,16 @@
+// Copyright 2019 The Operator-SDK Authors
```
This file isn't necessary, as everything in `test/` is for SDK testing and not exported.

xcoulon replied:
well, I had some errors reported by VSCode because there were no non-test go files in those packages, hence these 2 `doc.go` files, but I can remove them if you prefer.

xcoulon replied:
done in 2dfd90f

estroz commented on test/test-framework/test/e2e/doc.go:
```diff
@@ -0,0 +1,16 @@
+// Copyright 2019 The Operator-SDK Authors
```
This file isn't necessary, as everything in `test/` is for SDK testing and not exported.

xcoulon replied:
See comment above

xcoulon replied:
done in 2dfd90f

xcoulon commented:
@estroz I believe that I addressed all your comments. Please let me know if you have further concerns 

also, my apologies for the several pushes to fix the remaining compilation errors! I should have run `make install` well before! 😬

hasbro17 commented:
@xcoulon Thanks for the PR. I think this is pretty much in line with what we initially proposed on #890 but per my latest comment I think we still need to reach a consensus on some of the edge cases with these two flags before we can move forward with these changes.

xcoulon commented:
> @xcoulon Thanks for the PR. I think this is pretty much in line with what we initially proposed on #890 but per my latest comment I think we still need to reach a consensus on some of the edge cases with these two flags before we can move forward with these changes.

Thanks for your feedback, @hasbro17! I'll be happy to add more changes in this PR once the consensus is reached, in order to cover the edge cases ;)

LiliC commented:
@hasbro17 Do we still want this PR in?

hasbro17 commented:
Forgot to follow up on this. Thanks for reminding me @LiliC (and @openshift-ci-robot  😅 )

No, we probably don't want to configure the Operator or CRD scope from the CLI.
Per #890 and #1434 we're going to leave those changes to the user to keep the workflow simple.

@xcoulon Thanks for the PR and discussion but I'll close this for now.

xcoulon commented:
> Forgot to follow up on this. Thanks for reminding me @LiliC (and @openshift-ci-robot 😅 )
> 
> No, we probably don't want to configure the Operator or CRD scope from the CLI.
> Per #890 and #1434 we're going to leave those changes to the user to keep the workflow simple.
> 
> @xcoulon Thanks for the PR and discussion but I'll close this for now.

No worries, it was my pleasure to try to help here ;)
